cuda/matrix/hybrid_kernels.cu

```diff
@@ -0,0 +1,436 @@
+        for (auto i = threadIdx.x % warp_size; i < max_nnz_per_row;
+             i += warp_size) {
+            if (values[stride * i + row_idx] != zero<ValueType>()) {
+                part_result += 1;
+            }
+        }
+
+        auto warp_tile =
+            group::tiled_partition<warp_size>(group::this_thread_block());
+        result[row_idx] = reduce(
+            warp_tile, part_result,
+            [](const size_type &a, const size_type &b) { return a + b; });
+    }
+}
+
+
+// Copied from coo_kernel.cu
+template <typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void convert_coo_row_idxs_to_ptrs(
+        const IndexType *__restrict__ idxs, size_type num_nonzeros,
+        IndexType *__restrict__ ptrs, size_type length)
+{
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+
+    if (tidx == 0) {
+        ptrs[0] = 0;
+        ptrs[length - 1] = num_nonzeros;
+    }
+
+    if (0 < tidx && tidx < num_nonzeros) {
+        if (idxs[tidx - 1] < idxs[tidx]) {
+            for (auto i = idxs[tidx - 1] + 1; i <= idxs[tidx]; i++) {
+                ptrs[i] = tidx;
+            }
+        }
+    }
+}
+
+
+}  // namespace kernel
+
+
+namespace host_kernel {
+
+
+template <typename ValueType>
+ValueType calculate_nwarps(const size_type nnz, const ValueType nwarps_in_cuda)
+{
+    // TODO: multiple is a parameter should be tuned.
+    ValueType multiple = 8;
+    if (nnz >= 2000000) {
+        multiple = 128;
+    } else if (nnz >= 200000) {
+        multiple = 32;
+    }
+    return std::min(static_cast<int64>(multiple * nwarps_in_cuda),
+                    ceildiv(nnz, cuda_config::warp_size));
+}
+
+
+}  // namespace host_kernel
+
+
+template <typename ValueType, typename IndexType>
+void convert_to_csr(std::shared_ptr<const CudaExecutor> exec,
+                    matrix::Csr<ValueType, IndexType> *result,
+                    const matrix::Hybrid<ValueType, IndexType> *source)
+{
+    const auto num_rows = source->get_size()[0];
+    auto coo_offset = Array<IndexType>(exec, num_rows + 1);
+    auto coo_val = source->get_const_coo_values();
+    auto coo_col = source->get_const_coo_col_idxs();
+    auto coo_row = source->get_const_coo_row_idxs();
+    auto ell_val = source->get_const_ell_values();
+    auto ell_col = source->get_const_ell_col_idxs();
+    const auto stride = source->get_ell_stride();
+    const auto max_nnz_per_row = source->get_ell_num_stored_elements_per_row();
+    const auto coo_num_stored_elements = source->get_coo_num_stored_elements();
+
+    // Compute the row offset of Coo without zeros
+    size_type grid_num = ceildiv(coo_num_stored_elements, default_block_size);
+    kernel::convert_coo_row_idxs_to_ptrs<<<grid_num, default_block_size>>>(
+        as_cuda_type(coo_row), coo_num_stored_elements,
+        as_cuda_type(coo_offset.get_data()), num_rows + 1);
+
+    // Compute the row ptrs of Csr
+    using cuda_size_type = unsigned long long int;
```

Comment:
Please use something from `types.hpp`, or add it yourself and use it then, so we have all types in one place.
Also, this type, as you wrote it, is platform dependent, so please use either `size_t` (which is definitively also platform dependent, but in an understandable way), `uint64_t` (not platform dependent) or similar as a reference type.

It would be the easiest if you could just use `size_type` here, any reason why you did not use it?